cpp/modmesh/universe/rtree.hpp

```diff
@@ -0,0 +1,700 @@
+}; /* end of struct BoundingBox3D */
+
+/// Value operations traits for R-tree
+/// @tparam T Item type to be stored in R-tree
+/// @tparam BoundingBoxType Bounding box type associated with T
+template <typename T, typename BoundingBoxType>
+struct RTreeValueOps
+{
+    static BoundingBoxType get_bounding_box(T const & item);
+    static BoundingBoxType get_group_bounding_box(std::vector<T> const & items);
+}; /* end of struct RTreeValueOps */
+
+/// R-tree node structure
+/// @tparam T Item type to be stored in R-tree
+/// @tparam BoundingBoxType Bounding box type associated
+/// @tparam ValueOps Value operations traits for T and BoundingBoxType
+template <typename T, typename BoundingBoxType, typename ValueOpsType>
+struct RTreeNode
+{
+    BoundingBoxType bbox;
+    std::vector<T> items;
+
+    using RTreeNodeType = RTreeNode<T, BoundingBoxType, ValueOpsType>;
+    std::vector<std::unique_ptr<RTreeNodeType>> nodes;
+
+    RTreeNode(BoundingBoxType const & bbox_in)
+        : bbox(bbox_in)
+    {
+    }
+
+    ~RTreeNode() = default;
+    RTreeNode(RTreeNode const &) = delete;
+    RTreeNode & operator=(RTreeNode const &) = delete;
+    RTreeNode(RTreeNode &&) = delete;
+    RTreeNode & operator=(RTreeNode &&) = delete;
+
+    template <bool IS_LEAF>
+    auto & get_children()
+    {
+        if constexpr (IS_LEAF)
+        {
+            return items;
+        }
+        else
+        {
+            return nodes;
+        }
+    }
+
+    /// Recalculate bounding box based on contained items and nodes
+    void recalculate_bounding_box()
+    {
+        if (items.empty() && nodes.empty())
+        {
+            return; // empty node
+        }
+
+        bool first = true;
+        BoundingBoxType result_box = bbox;
+
+        // if leaf node
+        for (auto const & item : items)
+        {
+            BoundingBoxType item_box = ValueOpsType::get_bounding_box(item);
+            if (first)
+            {
+                result_box = item_box;
+                first = false;
+            }
+            else
+            {
+                result_box.expand(item_box);
+            }
+        }
+
+        // if internal node
+        for (auto const & child : nodes)
+        {
+            if (first)
+            {
+                result_box = child->bbox;
+                first = false;
+            }
+            else
+            {
+                result_box.expand(child->bbox);
+            }
+        }
+
+        bbox = result_box;
+    }
+}; /* end of struct RTreeNode */
+
+/// R-tree implementation for spatial index based on Guttman's R-tree paper
+/// @tparam T Item type to be stored in R-tree
+/// @tparam BoundingBoxType Bounding box type associated with T
+/// @tparam ValueOps Value operations traits for T and BoundingBoxType
+/// @tparam MAX_ITEMS_PER_NODE Maximum number of items per R-tree node
+template <typename T, typename BoundingBoxType, typename ValueOps = RTreeValueOps<T, BoundingBoxType>, int MAX_ITEMS_PER_NODE = 64>
```

Comment:
Rename `BoundingBoxType` as `B`.

Reply:
The real number value type should be named as `T`.  So the element type in the R-tree should be name as `E`.